### PR TITLE
Develop new

### PR DESCRIPTION
**Pull request by AadilLatif**

### CODE REVIEW

**Review by JKeen8 — CHANGES_REQUESTED** [review]
I ran these changes successfully in a cosim.   I had to delete hardcoded file exports in both motor models:
self.file = open(r"C:\Users\alatif\Desktop\NAERM\models\17735\motor.csv", "w")

File: PyDSS/helics_interface.py [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
-    n_states = 5
-    init_state = 1
+from PyDSS.pyContrReader import pyExportReader
+from PyDSS.simulation_input_models import SimulationSettingsModel
+from PyDSS.common import SUBSCRIPTIONS_FILENAME, ExportMode
+from PyDSS.pyLogger import getLoggerTag
+from PyDSS.utils.utils import load_data
 
-    type_info = {
+type_info = {
```
Comment (daniel-thom) [inline PyDSS/helics_interface.py:16]: Can you make this all-caps so that readers know it is a global “constant"?

File: PyDSS/helics_interface.py [hunk 2]
```diff
@@ -42,18 +43,153 @@ class helics_interface:
...
     }
+class DataType(Enum):
+    DOUBLE = "double"
+    VECTOR = "vector"
+    STRING = "string"
+    BOOLEAN = "boolean"
+    INTEGER = "integer"
+
+class Subscription(BaseModel):
+    model : str
```
Comment (daniel-thom) [inline PyDSS/helics_interface.py:54]: nit: typical formatting is `model: str`

**Comment by JKeen8** [comment]
@AadilLatif  as we discussed and as shown in the attachment, the Exports.toml file in the defaults directory doesn't follow the new exports toml format that you created for this branch.  
MicrosoftTeams-image (6)

File: PyDSS/helics_interface.py [hunk 3]
```diff
@@ -42,18 +43,153 @@ class helics_interface:
...
+    federate : Any
+    opendss_models : Dict
+    subscriptions : List[Subscription]
+
+    @validator('subscriptions')
+    def is_in_opendss_model(cls, v, values, **kwargs):
+        for subscription in v:
+            if subscription.model not in values["opendss_models"]:
+                raise ValidationError(f"The loaded OpenDSS model does not have an element define with the name {subscription}")
+            else:
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:86]: fixed

File: PyDSS/helics_interface.py [hunk 4]
```diff
@@ -42,18 +43,153 @@ class helics_interface:
         '%stored': 'double',
         'Distance': 'double'
     }
+class DataType(Enum):
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:46]: added

File: PyDSS/helics_interface.py [hunk 5]
```diff
@@ -42,18 +43,153 @@ class helics_interface:
...
+    object : Any = None
+    pub : Any = None
+    data_type: DataType
+    
+class Subscriptions(BaseModel):
+    federate : Any
+    opendss_models : Dict
+    subscriptions : List[Subscription]
+
+    @validator('subscriptions')
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:81]: fixed

File: PyDSS/helics_interface.py [hunk 6]
```diff
@@ -42,18 +43,153 @@ class helics_interface:
...
+        return v
+    
+class Publications(BaseModel):
+    
+    federate : Any
+    federate_name : str
+    opendss_models : Dict
+    publications : List[Publication] = []
+    legacy_input : Optional[Dict] = {}
+    input: Optional[Dict] = {}
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:103]: removed

File: PyDSS/pyControllers/Controllers/MotorStall.py [hunk 7]
```diff
@@ -79,6 +79,16 @@ def __init__(self, MotorObj, Settings, dssInstance, ElmObjectList, dssSolver):
...
         self.v = 0
+    
+        self._peak = 0
+        self.i2r = 0
+        
+        self.p_stall = 0
+        self.q_stall = 0
+        
+        self.file = open(r"C:\Users\alatif\Desktop\NAERM\models\17735\motor.csv", "w")
+        self.file.write("CompLF,CompPF,V_stall_adj,V_break_adj,P0,Q0,P,Q,u,x,theeta\n")
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/pyControllers/Controllers/MotorStall.py:90]: this was just for debugging purposes I will remove them,

File: examples/external_interfaces/Helics_example/run_dummy_federate.py.bak [hunk 8]
```diff
@@ -0,0 +1,71 @@
+import time
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply examples/external_interfaces/Helics_example/run_dummy_federate.py.bak:1]: removed

File: examples/external_interfaces/pydss_project/Exports/helics/CapacitorsInfo.csv [hunk 9]
```diff
@@ -0,0 +1 @@
+""
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply examples/external_interfaces/pydss_project/Exports/helics/CapacitorsInfo.csv:1]: removed

File: PyDSS/helics_interface.py (hunk 1 above)
Comment (AadilLatif) [reply PyDSS/helics_interface.py:16]: done

File: PyDSS/helics_interface.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:81]: fixed

**Comment by AadilLatif** [comment]
> @AadilLatif as we discussed and as shown in the attachment, the Exports.toml file in the defaults directory doesn't follow the new exports toml format that you created for this branch. MicrosoftTeams-image (6)

fixed

**Comment by AadilLatif** [comment]
> I ran these changes successfully in a cosim. I had to delete hardcoded file exports in both motor models: self.file = open(r"C:\Users\alatif\Desktop\NAERM\models\17735\motor.csv", "w")
> 
> Also, there are a lot print statements I'm getting that look like this that we should probably remove:
> 
> **PyDSS_5.Bus.1001458854_blv.puVmagAngle [0.7954920173250474, -121.48098708600301] PyDSS_5.Bus.1001458901_blv.puVmagAngle [0.7941299022136167, -121.67780292375414]**

fixed

File: PyDSS/helics_interface.py [hunk 10]
```diff
@@ -92,148 +228,128 @@ def _create_helics_federate(self):
...
+            
+            publication_dict = {
+                "opendss_models" : self._objects_by_class,
+                "federate_name" : self._settings.helics.federate_name,
+                "federate" : self._federate,
+            }
+            
+            if os.path.exists(legacy_export_file):
+                legacy_data = load_data(legacy_export_file)
+                print(legacy_data)
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py]: removed

File: PyDSS/helics_interface.py [hunk 11]
```diff
@@ -92,148 +228,128 @@ def _create_helics_federate(self):
...
+                publication_dict["legacy_input"] = legacy_data
+            elif os.path.exists(export_file):
```
[the remark this answers is not in the document]
Comment (AadilLatif) [reply PyDSS/helics_interface.py:313]: I guess they can. I would want to prioritize the new version.  Can I set priority for a validator

**Comment by AadilLatif** [comment]
> @AadilLatif as we discussed and as shown in the attachment, the Exports.toml file in the defaults directory doesn't follow the new exports toml format that you created for this branch. MicrosoftTeams-image (6)

fixed

**Review by JKeen8 — APPROVED** [review]

**Pull request merged**